src/framework/xr/xr-view.js

```diff
@@ -0,0 +1,333 @@
+     *
+     * - {@link XREYE_NONE}: None - inidcates a monoscopic view (likely mobile phone screen).
+     * - {@link XREYE_LEFT}: Left - indicates left eye view.
+     * - {@link XREYE_RIGHT}: Right - indicates a right eye view.
+     *
+     * @type {string}
+     */
+    get eye() {
+        return this._xrView.eye;
+    }
+
+    /**
+     * A Vec4 (x, y, width, height) that represents a view's viewport. For monoscopic screen
+     * it will define fullscreen view, but for stereoscopic views (left/right eye) it will define
+     * a part of a whole screen that view is occupying.
+     *
+     * @type {Vec4}
+     */
+    get viewport() {
+        return this._viewport;
+    }
+
+    /**
+     * @type {Mat4}
+     * @ignore
+     */
+    get projMat() {
+        return this._projMat;
+    }
+
+    /**
+     * @type {Mat4}
+     * @ignore
+     */
+    get projViewOffMat() {
+        return this._projViewOffMat;
+    }
+
+    /**
+     * @type {Mat4}
+     * @ignore
+     */
+    get viewOffMat() {
+        return this._viewOffMat;
+    }
+
+    /**
+     * @type {Mat4}
+     * @ignore
+     */
+    get viewInvOffMat() {
+        return this._viewInvOffMat;
+    }
+
+    /**
+     * @type {Mat3}
+     * @ignore
+     */
+    get viewMat3() {
+        return this._viewMat3;
+    }
+
+    /**
+     * @type {Float32Array}
+     * @ignore
+     */
+    get positionData() {
+        return this._positionData;
+    }
+
+    /**
+     * @param {*} frame - XRFrame from requestAnimationFrame callback.
+     * @param {XRView} xrView - XRView from WebXR API.
+     * @ignore
+     */
+    update(frame, xrView) {
+        this._xrView = xrView;
+        if (this._manager.views.availableColor)
+            this._xrCamera = this._xrView.camera;
+
+        const layer = frame.session.renderState.baseLayer;
+
+        // viewport
+        const viewport = layer.getViewport(this._xrView);
+        this._viewport.x = viewport.x;
+        this._viewport.y = viewport.y;
+        this._viewport.z = viewport.width;
+        this._viewport.w = viewport.height;
+
+        // matrices
+        this._projMat.set(this._xrView.projectionMatrix);
+        this._viewMat.set(this._xrView.transform.inverse.matrix);
+        this._viewInvMat.set(this._xrView.transform.matrix);
+
+        this._updateTextureColor();
+    }
+
+    /**
+     * @private
+     */
+    _updateTextureColor() {
+        if (!this._manager.views.availableColor || !this._xrCamera || !this._textureColor)
+            return;
+
+        const binding = this._manager.webglBinding;
+        if (!binding)
+            return;
+
+        const texture = binding.getCameraImage(this._xrCamera);
+        if (!texture)
+            return;
+
+        const device = this._manager.app.graphicsDevice;
+        const gl = device.gl;
+
+        if (!this._frameBufferSource) {
+            // create frame buffer to read from
+            this._frameBufferSource = gl.createFramebuffer();
```

Comment:
Last suggestion. I assume the device can be lost while this is active, and those framebuffers would no longer be valid / work, and so this should be handled.

the device fires this when the device is lost:
```           
 this.fire('devicelost');
```

So you should listen to this, and simply set gpu resources to null, so that they re-created the next frame, for example:

```
this._frameBufferSource = null;
```

It'd be good to test this by forcing device lost if possible on the device you're testing this on, to make sure nothing is missed here.

Reply:
I've improved device lost handling by camera, but apparently XrManager needs to do a better job for handling device lost also, and re-create XRWebGLLayer. I will make another PR for that.

Reply:
Implemented handling of WebGL context loss by WebXR here: https://github.com/playcanvas/engine/pull/5849